lib/api/apiUtils/bucket/bucketWebsite.js

```diff
@@ -0,0 +1,317 @@
+            return errors.InvalidArgument.customizeDescription(errMsg);
+        }
+        websiteConfig.setErrorDocument(parent[0].Key[0]);
+    }
+
+    if (parsingResult.RoutingRules) {
+        const routingRule = new RoutingRule();
+        const parent = parsingResult.RoutingRules;
+        if (!_xmlContainsElem(parent, 'RoutingRule', true)) {
+            errMsg = 'RoutingRules is not well-formed';
+            return errors.MalformedXML.customizeDescription(errMsg);
+        }
+        for (let i = 0; i < parent[0].RoutingRule.length; i++) {
+            const rule = parent[0].RoutingRule[i];
+            const redirectObj = {};
+            if (!(Array.isArray(rule.Redirect) && rule.Redirect.length === 1)) {
+                errMsg = 'RoutingRule requires Redirect, which is ' +
+                'missing or not well-formed';
+                return errors.MalformedXML.customizeDescription(errMsg);
+            }
+            // Looks like AWS doesn't actually make this check, but AWS
+            // documentation specifies at least one of the following elements
+            // must be in a Redirect rule. We also need at least one of the
+            // elements to know how to implement a redirect for a rule.
+            // http://docs.aws.amazon.com/AmazonS3/latest/API/
+            // RESTBucketPUTwebsite.html
+            if (!(_xmlContainsElem(rule.Redirect, 'Protocol')
+            || _xmlContainsElem(rule.Redirect, 'HostName')
+            || _xmlContainsElem(rule.Redirect, 'ReplaceKeyPrefixWith')
+            || _xmlContainsElem(rule.Redirect, 'ReplaceKeyWith')
+            || _xmlContainsElem(rule.Redirect, 'HttpRedirectCode'))) {
+                errMsg = 'Redirect must contain at least one of ' +
+                'following: Protocol, HostName, ReplaceKeyPrefixWith, ' +
+                'ReplaceKeyWith, or HttpRedirectCode element';
+                return errors.MalformedXML.customizeDescription(errMsg);
+            }
+            if (rule.Redirect[0].Protocol) {
+                if (rule.Redirect[0].Protocol[0] !== 'http' &&
+                rule.Redirect[0].Protocol[0] !== 'https') {
+                    errMsg = 'Invalid protocol, protocol can be http or ' +
+                    'https. If not defined, the protocol will be selected ' +
+                    'automatically.';
+                    return errors.InvalidRequest.customizeDescription(errMsg);
+                }
+                redirectObj.protocol = rule.Redirect[0].Protocol[0];
+            }
+            if (rule.Redirect[0].HttpRedirectCode) {
+                const code = parseInt(rule.Redirect[0].HttpRedirectCode[0], 10);
+                if (isNaN(code)) {
+                    errMsg = 'The provided HTTP redirect code is not valid. ' +
+                    'It should be a string containing a number.';
+                    return errors.MalformedXML.customizeDescription(errMsg);
+                }
+                if (!(code > 300 && code < 400)) {
```

Comment:
I tested AWS responses a bit, they return an error if the code is not in the right range:
<Error><Code>InvalidRequest</Code><Message>The provided HTTP redirect code (403) is not valid. Valid codes are 3XX except 300.</Message> [...] </Error>